File: base/macropod-python/rpc.py
from concurrent import futures
import os
import datetime
import grpc
import macropod_pb2 as pb
import macropod_pb2_grpc as pb_grpc
from google.protobuf.struct_pb2 import Struct
from google.protobuf import json_format

MAX_MESSAGE_LENGTH = 1024 * 1024 * 200
opts = [("grpc.max_receive_message_length", MAX_MESSAGE_LENGTH),("grpc.max_send_message_length", MAX_MESSAGE_LENGTH)]

def Timestamp(context, target, message):
    req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=context["Function"], WorkflowID=context["WorkflowID"], Depth=context["Depth"], Width=context["Width"], Target=target, Text=message)
    if "LOGGER" in os.environ:
        with grpc.insecure_channel(os.environ["LOGGER"], options=opts,) as channel:
            stub = pb_grpc.MacroPodLoggerStub(channel)
            stub.Timestamp(req)

def Error(context, err):
    req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=context["Function"], Text=err)
    if "LOGGER" in os.environ:
        with grpc.insecure_channel(os.environ["LOGGER"], options=opts,) as channel:
            stub = pb_grpc.MacroPodLoggerStub(channel)
            stub.Error(req)

def Print(context, message):
    req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=context["Function"], Text=message)
    if "LOGGER" in os.environ:
        with grpc.insecure_channel(os.environ["LOGGER"], options=opts,) as channel:
            stub = pb_grpc.MacroPodLoggerStub(channel)
            stub.Print(req)

def Invoke(context, dest, payload):
    Timestamp(context, dest, "Invoke")
    reply = ""
    code = 500
    target = os.environ[dest]
    if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
        target = os.environ["INGRESS"]
    req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=dest, Target=os.environ[dest], Text=payload, WorkflowID=context["WorkflowID"], Depth=(context["Depth"] + 1), Width=0)
    with grpc.insecure_channel(target, options=opts,) as channel:
        stub = pb_grpc.MacroPodFunctionStub(channel)
        if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
            stub = pb_grpc.MacroPodFunctionStub(channel)
            result = stub.Invoke(req)
            reply = result.Reply
            code = result.Code
        elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
            stub = pb_grpc.MacroPodIngressStub(channel)
            result = stub.FunctionInvoke(req)
            reply = result.Reply
            code = result.Code
        else:
            result = stub.Invoke(req)
            reply = result.Reply
            code = result.Code
    Timestamp(context, dest, "Invoke_End")
    return reply, code

def Invoke_JSON(context, dest, payload):
    Timestamp(context, dest, "Invoke_JSON")
    reply = ""
    code = 500
    target = os.environ[dest]
    if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
        target = os.environ["INGRESS"]
    payload_struct = Struct()
    payload_struct.update(payload)
    req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=dest, Target=os.environ[dest], JSON=payload_struct, WorkflowID=context["WorkflowID"], Depth=(context["Depth"] + 1), Width=0)
    with grpc.insecure_channel(target, options=opts,) as channel:
        stub = pb_grpc.MacroPodFunctionStub(channel)
        if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
            stub = pb_grpc.MacroPodFunctionStub(channel)
            result = stub.Invoke(req)
            reply = result.Reply
            code = result.Code
        elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
            stub = pb_grpc.MacroPodIngressStub(channel)
            result = stub.FunctionInvoke(req)
            reply = result.Reply
            code = result.Code
        else:
            result = stub.Invoke(req)
            reply = result.Reply
            code = result.Code
    Timestamp(context, dest, "Invoke_JSON_End")
    return reply, code

def Invoke_Data(context, dest, payload):
    Timestamp(context, dest, "Invoke_Data")
    reply = ""
    code = 500
    target = os.environ[dest]
    if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
        target = os.environ["INGRESS"]
    req = pb.MacroPodRequest(Data=payload, Workflow=context["Workflow"], Function=dest, Target=os.environ[dest], WorkflowID=context["WorkflowID"], Depth=(context["Depth"] + 1), Width=0)
    with grpc.insecure_channel(target, options=opts,) as channel:
        stub = pb_grpc.MacroPodFunctionStub(channel)
        if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
            stub = pb_grpc.MacroPodFunctionStub(channel)
            result = stub.Invoke(req)
            reply = result.Reply
            code = result.Code
        elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
            stub = pb_grpc.MacroPodIngressStub(channel)
            result = stub.FunctionInvoke(req)
            reply = result.Reply
            code = result.Code
        else:
            result = stub.Invoke(req)
            reply = result.Reply
            code = result.Code
    Timestamp(context, dest, "Invoke_Data_End")
    return reply, code

def Invoke_Multi(context, dest, payloads):
    Timestamp(context, dest, "Invoke_Multi")
    reply = []
    code = []
    target = os.environ[dest]
    if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
        target = os.environ["INGRESS"]
    with grpc.insecure_channel(target, options=opts,) as channel:
        stub = pb_grpc.MacroPodFunctionStub(channel)
        if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
            stub = pb_grpc.MacroPodFunctionStub(channel)
        elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
            stub = pb_grpc.MacroPodIngressStub(channel)
        tl = []
        with futures.ThreadPoolExecutor(max_workers=len(payloads)) as executor:
            for i in range(len(payloads)):
                req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=dest, Target=os.environ[dest], Text=payloads[i], WorkflowID=context["WorkflowID"], Depth=(context["Depth"] + 1), Width=i)
                if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
                    tl.append(executor.submit(stub.Invoke, req))
                elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
                    tl.append(executor.submit(stub.FunctionInvoke, req))
                else:
                    tl.append(executor.submit(stub.Invoke, req))
        reply = [t.result().Reply for t in tl]
        code = [t.result().Code for t in tl]
    Timestamp(context, dest, "Invoke_Multi_End")
    return reply, code

def Invoke_Multi_JSON(context, dest, payloads):
    Timestamp(context, dest, "Invoke_Multi_JSON")
    reply = []
    code = []
    target = os.environ[dest]
    if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
        target = os.environ["INGRESS"]
    with grpc.insecure_channel(target, options=opts,) as channel:
        stub = pb_grpc.MacroPodFunctionStub(channel)
        if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
            stub = pb_grpc.MacroPodFunctionStub(channel)
        elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
            stub = pb_grpc.MacroPodIngressStub(channel)
        tl = []
        with futures.ThreadPoolExecutor(max_workers=len(payloads)) as executor:
            for i in range(len(payloads)):
                payload_struct = Struct()
                payload_struct.update(payloads[i])
                req = pb.MacroPodRequest(Workflow=context["Workflow"], Function=dest, Target=os.environ[dest], JSON=payload_struct, WorkflowID=context["WorkflowID"], Depth=(context["Depth"] + 1), Width=i)
                if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
                    tl.append(executor.submit(stub.Invoke, req))
                elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
                    tl.append(executor.submit(stub.FunctionInvoke, req))
                else:
                    tl.append(executor.submit(stub.Invoke, req))
        reply = [t.result().Reply for t in tl]
        code = [t.result().Code for t in tl]
    Timestamp(context, dest, "Invoke_Multi_JSON_End")
    return reply, code

def Invoke_Multi_Data(context, dest, payloads):
    Timestamp(context, dest, "Invoke_Multi_Data")
    reply = []
    code = []
    target = os.environ[dest]
    if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
        target = os.environ["INGRESS"]
    with grpc.insecure_channel(target, options=opts,) as channel:
        stub = pb_grpc.MacroPodFunctionStub(channel)
        if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
            stub = pb_grpc.MacroPodFunctionStub(channel)
        elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
            stub = pb_grpc.MacroPodIngressStub(channel)
        tl = []
        with futures.ThreadPoolExecutor(max_workers=len(payloads)) as executor:
            for i in range(len(payloads)):
                req = pb.MacroPodRequest(Data=payloads[i], Workflow=context["Workflow"], Function=dest, Target=os.environ[dest], WorkflowID=context["WorkflowID"], Depth=(context["Depth"] + 1), Width=i)
                if "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "direct":
                    tl.append(executor.submit(stub.Invoke, req))
                elif "COMM_TYPE" in os.environ and os.environ["COMM_TYPE"] == "gateway":
                    tl.append(executor.submit(stub.FunctionInvoke, req))
                else:
                    tl.append(executor.submit(stub.Invoke, req))
        reply = [t.result().Reply for t in tl]
        code = [t.result().Code for t in tl]
    Timestamp(context, dest, "Invoke_Multi_Data_End")
    return reply, code

File: base/macropod-go/function/rpc.go
package function

import (
    pb "app/macropod_pb"

    "context"
    "time"
    "os"

    "google.golang.org/grpc"
    "google.golang.org/grpc/credentials/insecure"
    structpb "google.golang.org/protobuf/types/known/structpb"
)

type Context struct {
    Text       string
    JSON       map[string]interface{}
    Data       []byte
    Workflow   string
    Function   string
    WorkflowID string
    Depth      int32
    Width      int32
    Target     string
}

func Timestamp(ctx_in Context, target string, message string) {
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &ctx_in.Function, WorkflowID: &ctx_in.WorkflowID, Depth: &ctx_in.Depth, Width: &ctx_in.Width, Target: &target, Text: &message}
    logger := os.Getenv("LOGGER")
    if logger != "" {
        channel, _ := grpc.Dial(logger, grpc.WithDefaultCallOptions(), grpc.WithTransportCredentials(insecure.NewCredentials()))
        stub := pb.NewMacroPodLoggerClient(channel)
        stub.Timestamp(ctx, &req)
        channel.Close()
        cancel()
    }
}

func Error(ctx_in Context, err error) {
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    text := err.Error()
    req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &ctx_in.Function, Text: &text}
    logger := os.Getenv("LOGGER")
    if logger != "" {
        channel, _ := grpc.Dial(logger, grpc.WithDefaultCallOptions(), grpc.WithTransportCredentials(insecure.NewCredentials()))
        stub := pb.NewMacroPodLoggerClient(channel)
        stub.Error(ctx, &req)
        channel.Close()
        cancel()
    }
}

func Print(ctx_in Context, message string) {
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &ctx_in.Function, Text: &message}
    logger := os.Getenv("LOGGER")
    if logger != "" {
        channel, _ := grpc.Dial(logger, grpc.WithDefaultCallOptions(), grpc.WithTransportCredentials(insecure.NewCredentials()))
        stub := pb.NewMacroPodLoggerClient(channel)
        stub.Print(ctx, &req)
        channel.Close()
        cancel()
    }
}

func Invoke(ctx_in Context, dest string, payload string) (string, int32) {
    Timestamp(ctx_in, dest, "Invoke")
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    depth := ctx_in.Depth + 1
    width := int32(0)
    target := os.Getenv(dest)
    req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &dest, Target: &target, Text: &payload, WorkflowID: &ctx_in.WorkflowID, Depth: &depth, Width: &width}
    var res *pb.MacroPodReply
    var err error
    switch comm_type := os.Getenv("COMM_TYPE"); comm_type {
        case "direct":
            channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodFunctionClient(channel)
            res, err = stub.Invoke(ctx, &req)
            channel.Close()
        case "gateway":
            channel, _ := grpc.Dial(os.Getenv("INGRESS"), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodIngressClient(channel)
            res, err = stub.FunctionInvoke(ctx, &req)
            channel.Close()
        default:
            channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodFunctionClient(channel)
            res, err = stub.Invoke(ctx, &req)
            channel.Close()
    }
    cancel()
    if err != nil {
        Error(ctx_in, err)
    }
    Timestamp(ctx_in, dest, "Invoke_End")
    return res.GetReply(), res.GetCode()
}

func Invoke_JSON(ctx_in Context, dest string, payload map[string]interface{}) (string, int32) {
    Timestamp(ctx_in, dest, "Invoke_JSON")
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    depth := ctx_in.Depth + 1
    width := int32(0)
    j, _ := structpb.NewStruct(payload)
    target := os.Getenv(dest)
    req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &dest, Target: &target, JSON: j, WorkflowID: &ctx_in.WorkflowID, Depth: &depth, Width: &width}
    var res *pb.MacroPodReply
    var err error
    switch comm_type := os.Getenv("COMM_TYPE"); comm_type {
        case "direct":
            channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodFunctionClient(channel)
            res, err = stub.Invoke(ctx, &req)
            channel.Close()
        case "gateway":
            channel, _ := grpc.Dial(os.Getenv("INGRESS"), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodIngressClient(channel)
            res, err = stub.FunctionInvoke(ctx, &req)
            channel.Close()
        default:
            channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodFunctionClient(channel)
            res, err = stub.Invoke(ctx, &req)
            channel.Close()
    }
    cancel()
    if err != nil {
        Error(ctx_in, err)
    }
    Timestamp(ctx_in, dest, "Invoke_JSON_End")
    return res.GetReply(), res.GetCode()
}

func Invoke_Data(ctx_in Context, dest string, payload []byte) (string, int32) {
    Timestamp(ctx_in, dest, "Invoke_Data")
    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    depth := ctx_in.Depth + 1
    width := int32(0)
    target := os.Getenv(dest)
    req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &dest, Target: &target, Data: payload, WorkflowID: &ctx_in.WorkflowID, Depth: &depth, Width: &width}
    var res *pb.MacroPodReply
    var err error
    switch comm_type := os.Getenv("COMM_TYPE"); comm_type {
        case "direct":
            channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodFunctionClient(channel)
            res, err = stub.Invoke(ctx, &req)
            channel.Close()
        case "gateway":
            channel, _ := grpc.Dial(os.Getenv("INGRESS"), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodIngressClient(channel)
            res, err = stub.FunctionInvoke(ctx, &req)
            channel.Close()
        default:
            channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
            stub := pb.NewMacroPodFunctionClient(channel)
            res, err = stub.Invoke(ctx, &req)
            channel.Close()
    }
    cancel()
    if err != nil {
        Error(ctx_in, err)
    }
    Timestamp(ctx_in, dest, "Invoke_Data_End")
    return res.GetReply(), res.GetCode()
}

func Invoke_Multi(ctx_in Context, dest string, payloads []string) ([]string, []int32) {
    Timestamp(ctx_in, dest, "Invoke_Multi")
    tl := make(chan *pb.MacroPodReply)
    for i := 0; i < len(payloads); i++ {
        go func(n int) {
            ctx, cancel := context.WithTimeout(context.Background(), time.Second)
            depth := ctx_in.Depth + 1
            width := int32(n)
            target := os.Getenv(dest)
            req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &dest, Target: &target, Text: &payloads[n], WorkflowID: &ctx_in.WorkflowID, Depth: &depth, Width: &width}
            var res *pb.MacroPodReply
            var err error
            switch comm_type := os.Getenv("COMM_TYPE"); comm_type {
                case "direct":
                    channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodFunctionClient(channel)
                    res, err = stub.Invoke(ctx, &req)
                    channel.Close()
                case "gateway":
                    channel, _ := grpc.Dial(os.Getenv("INGRESS"), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodIngressClient(channel)
                    res, err = stub.FunctionInvoke(ctx, &req)
                    channel.Close()
                default:
                    channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodFunctionClient(channel)
                    res, err = stub.Invoke(ctx, &req)
                    channel.Close()
            }
            if err != nil {
                Error(ctx_in, err)
            }
            tl <- res
            cancel()
        }(i)
    }
    reply := make([]string, 0)
    code := make([]int32, 0)
    for i := 0; i < len(payloads); i++ {
        res := <-tl
        reply = append(reply, res.GetReply())
        code = append(code, res.GetCode())
    }
    Timestamp(ctx_in, dest, "Invoke_Multi_End")
    return reply, code
}

func Invoke_Multi_JSON(ctx_in Context, dest string, payloads []map[string]interface{}) ([]string, []int32) {
    Timestamp(ctx_in, dest, "Invoke_Multi_JSON")
    tl := make(chan *pb.MacroPodReply)
    for i := 0; i < len(payloads); i++ {
        go func(n int) {
            ctx, cancel := context.WithTimeout(context.Background(), time.Second)
            depth := ctx_in.Depth + 1
            width := int32(n)
            j, _ := structpb.NewStruct(payloads[n])
            target := os.Getenv(dest)
            req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &dest, Target: &target, JSON: j, WorkflowID: &ctx_in.WorkflowID, Depth: &depth, Width: &width}
            var res *pb.MacroPodReply
            var err error
            switch comm_type := os.Getenv("COMM_TYPE"); comm_type {
                case "direct":
                    channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodFunctionClient(channel)
                    res, err = stub.Invoke(ctx, &req)
                    channel.Close()
                case "gateway":
                    channel, _ := grpc.Dial(os.Getenv("INGRESS"), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodIngressClient(channel)
                    res, err = stub.FunctionInvoke(ctx, &req)
                    channel.Close()
                default:
                    channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodFunctionClient(channel)
                    res, err = stub.Invoke(ctx, &req)
                    channel.Close()
            }
            if err != nil {
                Error(ctx_in, err)
            }
            tl <- res
            cancel()
        }(i)
    }
    reply := make([]string, 0)
    code := make([]int32, 0)
    for i := 0; i < len(payloads); i++ {
        res := <-tl
        reply = append(reply, res.GetReply())
        code = append(code, res.GetCode())
    }
    Timestamp(ctx_in, dest, "Invoke_Multi_JSON_End")
    return reply, code
}

func Invoke_Multi_Data(ctx_in Context, dest string, payloads [][]byte) ([]string, []int32) {
    Timestamp(ctx_in, dest, "Invoke_Multi_Data")
    tl := make(chan *pb.MacroPodReply)
    for i := 0; i < len(payloads); i++ {
        go func(n int) {
            ctx, cancel := context.WithTimeout(context.Background(), time.Second)
            depth := ctx_in.Depth + 1
            width := int32(n)
            target := os.Getenv(dest)
            req := pb.MacroPodRequest{Workflow: &ctx_in.Workflow, Function: &dest, Target: &target, Data: payloads[n], WorkflowID: &ctx_in.WorkflowID, Depth: &depth, Width: &width}
            var res *pb.MacroPodReply
            var err error
            switch comm_type := os.Getenv("COMM_TYPE"); comm_type {
                case "direct":
                    channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodFunctionClient(channel)
                    res, err = stub.Invoke(ctx, &req)
                    channel.Close()
                case "gateway":
                    channel, _ := grpc.Dial(os.Getenv("INGRESS"), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodIngressClient(channel)
                    res, err = stub.FunctionInvoke(ctx, &req)
                    channel.Close()
                default:
                    channel, _ := grpc.Dial(target, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1024*1024*200), grpc.MaxCallSendMsgSize(1024*1024*200)), grpc.WithTransportCredentials(insecure.NewCredentials()))
                    stub := pb.NewMacroPodFunctionClient(channel)
                    res, err = stub.Invoke(ctx, &req)
                    channel.Close()
            }
            if err != nil {
                Error(ctx_in, err)
            }
            tl <- res
            cancel()
        }(i)
    }
    reply := make([]string, 0)
    code := make([]int32, 0)
    for i := 0; i < len(payloads); i++ {
        res := <-tl
        reply = append(reply, res.GetReply())
        code = append(code, res.GetCode())
    }
    Timestamp(ctx_in, dest, "Invoke_Multi_Data_End")
    return reply, code
}
